GPS.device.nut

```diff
@@ -0,0 +1,258 @@
+                    
+                    retTable.status <- "Active";
+                    retTable.fixQuality <- parsedFields[6];
+                    retTable.numSatellites <- parsedFields[7].tointeger();
+                    retTable.altitude <- parsedFields[9].tofloat();
+                    break;
+                // Satellites in view
+                case GPS_GSV:
+                    retTable.type <- GPS_GSV;
+                    retTable.numSatellites <- parsedFields[3].tointeger();
+                    break;
+                // GPS DOP and active satellites
+                case GPS_GSA:
+                    retTable.type <- GPS_GSA;
+                    retTable.threeDFix <- parsedFields[2];
+                    local len = parsedFields.len();
+                    // need to use the len b/c we don't know the number of satellite PRNs
+                    retTable.PDOP <- parsedFields[len-4];
+                    retTable.HDOP <- parsedFields[len-3];
+                    retTable.VDOP <- parsedFields[len-2];
+                    break;
+                // return empty table if not a supported data type
+                default:
+                    return {};
+            }
+        }
+        local checkLen = parsedFields[parsedFields.len() - 1].len();
+        retTable.checkSum <- parsedFields[parsedFields.len()-1].slice(checkLen-4, checkLen-2);
+        return retTable;
+    }
+
+    // Get the latitude from strings. Store it in tb
+    function _extractLat(str, tb, direction) {
+        local lat = str.slice(0, 2).tofloat() + (str.slice(2).tofloat()/60);
+        if (direction == "S") lat = -lat;
+        tb.latitude <- lat;
+    }
+
+    // Get the longitude from strings. Store it in tb
+    function _extractLong(str, tb, direction) {
+        local long = str.slice(0, 3).tofloat() + (str.slice(3).tofloat()/60);
+        if (direction == "W") long = -long;
+        tb.longitude <- long;
+    }
+
+    // Get the time from a string. Store it in tb
+    function _extractTime(str, tb) {
+        local time = str.tointeger();
+                    
+        tb.seconds <- time%100;
+        time = time/100;
+        tb.minutes <- time%100;
+        time = time/100;
+        tb.hours <- time; 
+    }
+    
+}
+
+
+
+class GPS {
+    
+    static VERSION = "1.0.0";
+
+    _gpsLine = "";
+    gpsCounter = 0;
+    gpsRate = 30;
+    _fields = null;
+    _gps = null;
+    _lastTable = null;
+    _lastLat = 0;
+    _lastLong = 0;
+    _isValid = false;
+    _numSatellites = 0;
+    _fix = false;
+    _fixCallback = null;
+
+    static LINE_MAX = 150;
+    
+    constructor(uart, fixCallback, baudrate=9600) {
+        _gps = uart;
+        _fields = GPSFields();
+        // GPS is configured by the constructor so that we can register
+        // the rxdata callback
+        _gps.configure(baudrate, 8, PARITY_NONE, 1, NO_CTSRTS, _gpsRxdata.bindenv(this));
+        _fixCallback = fixCallback;
+    }
+
+    function getLastLatitude() {
```

Comment:
These getLastLatitude/Longitude would be better if they included a the time this info was updated. It might be better to have a single getLastLocation function that returns the latest latitude, longitude and time the data was collected.